File: wimg-ios/wimg/Services/DemoDataService.swift
import Foundation

struct DemoDataService {
    private static let demoLoadedKey = "wimg_demo_loaded"

    // MARK: - Fixed monthly transactions

    private static let fixedMonthly: [(desc: String, amount: Int?)] = [
        ("GEHALT {MONTH} 2026 ARBEITGEBER GMBH", 325_000),
        ("MIETE {MONTH} 2026 HAUSVERWALTUNG", -95_000),
        ("STADTWERKE STROM GAS", nil), // random -9500...-11500
        ("NETFLIX.COM", -1_799),
        ("SPOTIFY AB", -999),
        ("ALLIANZ VERSICHERUNG", -8_950),
        ("GEZ BEITRAGSSERVICE", -1_836),
        ("VODAFONE GMBH MOBILFUNK", -3_999),
    ]

    private static let frequent: [(desc: String, min: Int, max: Int, freqMin: Int, freqMax: Int)] = [
        ("REWE SAGT DANKE {id}//MUENCHEN/DE", -8_500, -1_500, 3, 4),
        ("LIDL DIENSTL SAGT DANKE", -4_500, -1_200, 2, 3),
        ("EDEKA CENTER {id}", -5_500, -800, 2, 3),
        ("DM DROGERIEMARKT SAGT DANKE", -2_500, -500, 1, 2),
        ("DB VERTRIEB GMBH", -4_500, -1_500, 1, 2),
    ]

    private static let occasional: [(desc: String, min: Int, max: Int)] = [
        ("LIEFERANDO.DE", -3_500, -1_200),
        ("AMAZON EU SARL", -12_000, -1_500),
        ("ROSSMANN SAGT DANKE", -2_000, -500),
        ("APOTHEKE AM MARKT", -3_000, -500),
    ]

    private static let monthNames = [
        "", "JANUAR", "FEBRUAR", "MAERZ", "APRIL", "MAI", "JUNI",
        "JULI", "AUGUST", "SEPTEMBER", "OKTOBER", "NOVEMBER", "DEZEMBER",
    ]

    // MARK: - Public API

    static var isDemoLoaded: Bool {
        UserDefaults.standard.bool(forKey: demoLoadedKey)
    }

    static func clearDemoFlag() {
        UserDefaults.standard.removeObject(forKey: demoLoadedKey)
    }

    static func loadDemoData() {
        let csv = generateDemoCSV()
        guard let data = csv.data(using: .isoLatin1) else { return }

        do {
            let result = try LibWimg.importCSV(data)
            if result.imported > 0 {
                _ = LibWimg.autoCategorize()
                UserDefaults.standard.set(true, forKey: demoLoadedKey)
                NotificationCenter.default.post(name: .wimgDataChanged, object: nil)
            }
        } catch {
            print("[wimg] Demo import failed: \(error)")
        }
    }

    // MARK: - CSV Generation

    static func generateDemoCSV() -> String {
        let calendar = Calendar.current
        let now = Date()
        var rows: [(date: String, desc: String, amount: String, sortKey: Int)] = []

        for offset in 0..<3 {
            guard let monthDate = calendar.date(byAdding: .month, value: -offset, to: now) else { continue }
            let year = calendar.component(.year, from: monthDate)
            let month = calendar.component(.month, from: monthDate)
            let maxDay = calendar.range(of: .day, in: .month, for: monthDate)?.count ?? 28
            let monthName = monthNames[month]

            // Fixed monthly (day 1-5)
            for tx in fixedMonthly {
                let day = Int.random(in: 1...min(5, maxDay))
                let desc = tx.desc.replacingOccurrences(of: "{MONTH}", with: monthName)
                let cents: Int
                if let fixed = tx.amount {
                    cents = fixed
                } else {
                    cents = Int.random(in: -11_500 ... -9_500)
                }
                rows.append((
                    date: formatDate(year: year, month: month, day: day),
                    desc: desc,
                    amount: formatAmount(cents: cents),
                    sortKey: year * 10000 + month * 100 + day
                ))
            }

            // Frequent
            for tx in frequent {
                let count = Int.random(in: tx.freqMin...tx.freqMax)
                for _ in 0..<count {
                    let day = Int.random(in: 1...maxDay)
                    let id = String(Int.random(in: 10000...99999))
                    let desc = tx.desc.replacingOccurrences(of: "{id}", with: id)
                    let cents = Int.random(in: tx.min...tx.max)
                    rows.append((
                        date: formatDate(year: year, month: month, day: day),
                        desc: desc,
                        amount: formatAmount(cents: cents),
                        sortKey: year * 10000 + month * 100 + day
                    ))
                }
            }

            // Occasional (0-2x)
            for tx in occasional {
                let count = Int.random(in: 0...2)
                for _ in 0..<count {
                    let day = Int.random(in: 5...maxDay)
                    let cents = Int.random(in: tx.min...tx.max)
                    rows.append((
                        date: formatDate(year: year, month: month, day: day),
                        desc: tx.desc,
                        amount: formatAmount(cents: cents),
                        sortKey: year * 10000 + month * 100 + day
                    ))
                }
            }
        }

        // Sort descending (Comdirect order)
        rows.sort { $0.sortKey > $1.sortKey }

        let header = "\"Buchungstag\";\"Wertstellung (Valuta)\";\"Vorgang\";\"Buchungstext\";\"Umsatz in EUR\""
        let lines = rows.map { r in
            "\"\(r.date)\";\"\(r.date)\";\"Lastschrift\";\"\(r.desc)\";\"\(r.amount)\""
        }

        return header + "\n" + lines.joined(separator: "\n") + "\n"
    }

    // MARK: - Helpers

    private static func formatDate(year: Int, month: Int, day: Int) -> String {
        String(format: "%02d.%02d.%04d", day, month, year)
    }

    private static func formatAmount(cents: Int) -> String {
        let sign = cents < 0 ? "-" : ""
        let abs = Swift.abs(cents)
        let eur = abs / 100
        let ct = abs % 100

        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.locale = Locale(identifier: "de_DE")
        formatter.groupingSeparator = "."
        let eurStr = formatter.string(from: NSNumber(value: eur)) ?? "\(eur)"

        return "\(sign)\(eurStr),\(String(format: "%02d", ct))"
    }
}

File: wimg-android/app/src/main/java/com/wimg/app/services/DemoDataService.kt
package com.wimg.app.services

import android.content.Context
import com.wimg.app.bridge.LibWimg
import java.text.NumberFormat
import java.util.Calendar
import java.util.Locale
import kotlin.math.abs
import kotlin.random.Random

object DemoDataService {
    private const val DEMO_LOADED_KEY = "wimg_demo_loaded"

    private data class FixedTx(val desc: String, val amount: Int?)
    private data class FrequentTx(val desc: String, val min: Int, val max: Int, val freqMin: Int, val freqMax: Int)
    private data class OccasionalTx(val desc: String, val min: Int, val max: Int)

    private val fixedMonthly = listOf(
        FixedTx("GEHALT {MONTH} 2026 ARBEITGEBER GMBH", 325_000),
        FixedTx("MIETE {MONTH} 2026 HAUSVERWALTUNG", -95_000),
        FixedTx("STADTWERKE STROM GAS", null),
        FixedTx("NETFLIX.COM", -1_799),
        FixedTx("SPOTIFY AB", -999),
        FixedTx("ALLIANZ VERSICHERUNG", -8_950),
        FixedTx("GEZ BEITRAGSSERVICE", -1_836),
        FixedTx("VODAFONE GMBH MOBILFUNK", -3_999),
    )

    private val frequent = listOf(
        FrequentTx("REWE SAGT DANKE {id}//MUENCHEN/DE", -8_500, -1_500, 3, 4),
        FrequentTx("LIDL DIENSTL SAGT DANKE", -4_500, -1_200, 2, 3),
        FrequentTx("EDEKA CENTER {id}", -5_500, -800, 2, 3),
        FrequentTx("DM DROGERIEMARKT SAGT DANKE", -2_500, -500, 1, 2),
        FrequentTx("DB VERTRIEB GMBH", -4_500, -1_500, 1, 2),
    )

    private val occasional = listOf(
        OccasionalTx("LIEFERANDO.DE", -3_500, -1_200),
        OccasionalTx("AMAZON EU SARL", -12_000, -1_500),
        OccasionalTx("ROSSMANN SAGT DANKE", -2_000, -500),
        OccasionalTx("APOTHEKE AM MARKT", -3_000, -500),
    )

    private val monthNames = arrayOf(
        "", "JANUAR", "FEBRUAR", "MAERZ", "APRIL", "MAI", "JUNI",
        "JULI", "AUGUST", "SEPTEMBER", "OKTOBER", "NOVEMBER", "DEZEMBER",
    )

    fun isDemoLoaded(context: Context): Boolean {
        return context.getSharedPreferences("wimg", 0).getBoolean(DEMO_LOADED_KEY, false)
    }

    fun loadDemoData(context: Context) {
        val csv = generateDemoCSV()
        val data = csv.toByteArray(Charsets.ISO_8859_1)
        val result = LibWimg.importCsv(data)
        if (result != null && result.imported > 0) {
            LibWimg.autoCategorize()
            context.getSharedPreferences("wimg", 0).edit().putBoolean(DEMO_LOADED_KEY, true).apply()
        }
    }

    private fun generateDemoCSV(): String {
        val cal = Calendar.getInstance()
        data class Row(val date: String, val desc: String, val amount: String, val sortKey: Int)
        val rows = mutableListOf<Row>()

        for (offset in 0 until 3) {
            cal.time = java.util.Date()
            cal.add(Calendar.MONTH, -offset)
            val year = cal.get(Calendar.YEAR)
            val month = cal.get(Calendar.MONTH) + 1
            val maxDay = cal.getActualMaximum(Calendar.DAY_OF_MONTH)
            val monthName = monthNames[month]

            // Fixed monthly
            for (tx in fixedMonthly) {
                val day = Random.nextInt(1, minOf(6, maxDay + 1))
                val desc = tx.desc.replace("{MONTH}", monthName)
                val cents = tx.amount ?: Random.nextInt(-11_500, -9_499)
                rows.add(Row(formatDate(year, month, day), desc, formatAmount(cents), year * 10000 + month * 100 + day))
            }

            // Frequent
            for (tx in frequent) {
                val count = Random.nextInt(tx.freqMin, tx.freqMax + 1)
                repeat(count) {
                    val day = Random.nextInt(1, maxDay + 1)
                    val id = Random.nextInt(10000, 100000).toString()
                    val desc = tx.desc.replace("{id}", id)
                    val cents = Random.nextInt(tx.min, tx.max + 1)
                    rows.add(Row(formatDate(year, month, day), desc, formatAmount(cents), year * 10000 + month * 100 + day))
                }
            }

            // Occasional
            for (tx in occasional) {
                val count = Random.nextInt(0, 3)
                repeat(count) {
                    val day = Random.nextInt(5, maxDay + 1)
                    val cents = Random.nextInt(tx.min, tx.max + 1)
                    rows.add(Row(formatDate(year, month, day), tx.desc, formatAmount(cents), year * 10000 + month * 100 + day))
                }
            }
        }

        rows.sortByDescending { it.sortKey }

        val header = "\"Buchungstag\";\"Wertstellung (Valuta)\";\"Vorgang\";\"Buchungstext\";\"Umsatz in EUR\""
        val lines = rows.joinToString("\n") { r ->
            "\"${r.date}\";\"${r.date}\";\"Lastschrift\";\"${r.desc}\";\"${r.amount}\""
        }
        return "$header\n$lines\n"
    }

    private fun formatDate(year: Int, month: Int, day: Int): String {
        return String.format("%02d.%02d.%04d", day, month, year)
    }

    private fun formatAmount(cents: Int): String {
        val sign = if (cents < 0) "-" else ""
        val absCents = abs(cents)
        val eur = absCents / 100
        val ct = absCents % 100
        val nf = NumberFormat.getIntegerInstance(Locale.GERMANY)
        return "$sign${nf.format(eur)},${String.format("%02d", ct)}"
    }
}
